crates/sui-core/src/overload_monitor.rs

```diff
@@ -0,0 +1,292 @@
+    let execution_rate = authority.metrics.execution_rate_tracker.lock().rate();
+
+    let (is_overload, load_shedding_percentage) =
+        check_overload_signals(config, queueing_latency, txn_ready_rate, execution_rate);
+    if is_overload {
+        authority
+            .overload_info
+            .set_overload(load_shedding_percentage);
+    } else {
+        authority.overload_info.clear_overload();
+    }
+
+    authority
+        .metrics
+        .authority_overload_status
+        .set(is_overload as i64);
+    authority
+        .metrics
+        .authority_load_shedding_percentage
+        .set(load_shedding_percentage as i64);
+    true
+}
+
+// Calculates the percentage of transactions to drop in order to reduce execution queue.
+// Returns the integer percentage between 0 and 100.
+fn calculate_load_shedding_percentage(txn_ready_rate: f64, execution_rate: f64) -> u32 {
+    // When transaction ready rate is practically 0, we aren't adding more load to the
+    // execution driver, so no shedding.
+    // TODO: consensus handler or transaction manager can also be overloaded.
+    if txn_ready_rate < 1e-10 {
+        return 0;
+    }
+
+    // Deflate the execution rate to account for the case that execution_rate is close to
+    // txn_ready_rate.
+    if execution_rate * 0.9 > txn_ready_rate {
+        return 0;
+    }
+
+    // In order to maintain execution queue length, we need to drop at least (1 - executionRate / readyRate).
+    // TO reduce the queue length, here we add 10% more transactions to drop.
+    (((1.0 - execution_rate * 0.9 / txn_ready_rate) + 0.1).min(1.0) * 100.0).round() as u32
+}
+
+// Given overload signals (`queueing_latency`, `txn_ready_rate`, `execution_rate`), return whether
+// the authority server should enter load shedding mode, and how much percentage of transactions to drop.
+fn check_overload_signals(
+    config: &OverloadThresholdConfig,
+    queueing_latency: Duration,
+    txn_ready_rate: f64,
+    execution_rate: f64,
+) -> (bool, u32) {
+    let overload_status;
+    let load_shedding_percentage;
+    if queueing_latency > config.execution_queue_latency_hard_limit {
+        overload_status = true;
+        load_shedding_percentage = max(
+            calculate_load_shedding_percentage(txn_ready_rate, execution_rate),
+            config.min_load_shedding_percentage_above_hard_limit,
+        );
+    } else if queueing_latency > config.execution_queue_latency_soft_limit {
+        load_shedding_percentage =
+            calculate_load_shedding_percentage(txn_ready_rate, execution_rate);
+        overload_status = load_shedding_percentage > 0;
+    } else {
+        overload_status = false;
+        load_shedding_percentage = 0;
+    }
+
+    let load_shedding_percentage = min(
+        load_shedding_percentage,
+        config.max_load_shedding_percentage,
+    );
+    (overload_status, load_shedding_percentage)
+}
+
+#[cfg(test)]
```

Comment:
I think we need a simulation test of some sort here. I don't mean an e2e simulator test, just a test with:

- An event emitter (aka "load generator") that generates events at a rate that varies over time
- A load shedder that sheds load according to the logic in this module
- An event consumer (aka "executor") that consumes events at some rate.

Then we can get some idea of what happens in the following scenarios
- load spike
- execution slow down
- rapid load oscillations / spikes
- very low traffic rates

The main thing we would want to see is whether the system produces oscillations on its own, and whether it can return to normal quickly after an adverse event is over.

But, this is just a suggestion, if you have a quicker / easier way to evaluate this let me know. I just think it will be very time consuming to do these tests in a test cluster with a load generator. (We'll need to do that as well to fully validate this, but that will have a much slower feedback loop and we may not be able to test as many scenarios)

. Otherwise I don't know how to begin reasoning about what kind of behavior this system might lead to. For instance

Basically set up a scenario with a randomized load generator and

Reply:
100% agree!

My plan was to write such test after I finish the actual shedding of the transaction. But I think I can create one that checks system status. Later on when shedding is implemented, I can evolve that test to further see the effectiveness of shedding.

Let me do it in a separate PR.

Reply:
sounds good